### PR TITLE
Vmal

### PR DESCRIPTION
**Pull request by Malisevych**

### CODE REVIEW

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
+package com.softserve.if072.mvcapp.service;
+
+import com.softserve.if072.common.model.*;
```
Comment (kucherif) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java]: .

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/controller/ProductPageController.java [hunk 2]
```diff
@@ -35,19 +38,14 @@
 
 @Controller
 @RequestMapping("/product")
+@PropertySource(value = {"classpath:application.properties", "classpath:message.properties"})
 public class ProductPageController extends BaseController {
 
-    @Value("${application.restProductURL}")
-    private String productUrl;
+    @Autowired
+    private ProductPageService productPageService;
```
Comment (kucherif) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/controller/ProductPageController.java]: do autowire via constructor

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+    private String unitUrl;
+
+    @Value("${application.restCategoryURL}")
+    private String categoryUrl;
+
+    @Value("${application.restStoreURL}")
+    private String storeUrl;
+
+    @Autowired
+    RestTemplate restTemplate;
```
Comment (kucherif) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java]: Access modifier is missed

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+    @Value("${application.restUnitURL}")
+    private String unitUrl;
+
+    @Value("${application.restCategoryURL}")
+    private String categoryUrl;
+
+    @Value("${application.restStoreURL}")
+    private String storeUrl;
+
+    @Autowired
```
Comment (kucherif) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java]: autowire via constructor

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+    public boolean isAlreadyExist(Product product, User user) {
+
+        Product existsProduct = getProductByNameAndUserId(product, user);
+
+        if(existsProduct != null && existsProduct.isEnabled()) {
+            if(existsProduct.getId() == product.getId()) {
+                return false;
+            } else {
+                return true;
+            }
```
Comment (kucherif) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java]:
```
 if(existsProduct.getId() == product.getId()) {
                return false;
            } else {
                return true;
            }
```

can be replaced with 
`return !existsProduct.getId() == product.getId()`

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+
+    public boolean isDeleted(Product product, User user) {
+
+        Product existsProduct = getProductByNameAndUserId(product, user);
+
+        if(existsProduct != null && !existsProduct.isEnabled()) {
+            return true;
+        } else {
+            return false;
+        }
```
Comment (kucherif) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java]: the same as above

File: MVCapplication/src/main/webapp/WEB-INF/views/product.jsp [hunk 7]
```diff
@@ -12,6 +12,9 @@
 <style>
     #image, #edit,#delete, #goStores, #addImage, #addImage, #addToShoppingList {text-align:center; cursor: pointer}
     #image {text-align:center}
+
+    td {vertical-align: middle}
+    .fa:hover {color: #7B83EF}
```
Comment (kucherif) [inline MVCapplication/src/main/webapp/WEB-INF/views/product.jsp]: move to separate css file

File: RESTserver/src/main/java/com/softserve/if072/restservice/controller/ProductController.java [hunk 8]
```diff
@@ -11,16 +11,7 @@
...
-import org.springframework.web.bind.annotation.GetMapping;
-import org.springframework.web.bind.annotation.PathVariable;
-import org.springframework.web.bind.annotation.PostMapping;
-import org.springframework.web.bind.annotation.PutMapping;
-import org.springframework.web.bind.annotation.RequestBody;
-import org.springframework.web.bind.annotation.RequestMapping;
-import org.springframework.web.bind.annotation.ResponseBody;
-import org.springframework.web.bind.annotation.ResponseStatus;
-import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.bind.annotation.*;
```
Comment (kucherif) [inline RESTserver/src/main/java/com/softserve/if072/restservice/controller/ProductController.java]: .

File: RESTserver/src/main/java/com/softserve/if072/restservice/service/security/ProductSecurityService.java [hunk 9]
```diff
@@ -21,4 +21,20 @@ public boolean hasPermissionToAccess(int productId){
...
+        Product product = productDAO.getProductByName(productName);
+
+        if(product == null) {
+            return true;
+        } else {
+            if(product != null && product.getUser() != null && product.getUser().getId() == getCurrentUser().getId()) {
+                return true;
+            } else {
+                return false;
+            }
```
Comment (kucherif) [inline RESTserver/src/main/java/com/softserve/if072/restservice/service/security/ProductSecurityService.java]:
```
            if(product != null && product.getUser() != null && product.getUser().getId() == getCurrentUser().getId()) {
                return true;
            } else {
                return false;
            }
```
Can be replaced with
`return product != null && product.getUser() != null && product.getUser().getId() == getCurrentUser().getId()`

File: common-modules/src/main/java/com/softserve/if072/common/model/validation/product/ValidUnit.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * The annotation for validation of selected units
+ *
+ * @author Vitaliy Malisevych
+ */
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = UnitValidator.class)
+
+public @interface ValidUnit {
```
Comment (kucherif) [inline common-modules/src/main/java/com/softserve/if072/common/model/validation/product/ValidUnit.java]: 👍 for custom validation

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/controller/ProductPageController.java (hunk 2 above)
Comment (OlehPochernin) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/controller/ProductPageController.java]: couldn't you explain why autowiring via constructor is better?

Comment (kucherif) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/controller/ProductPageController.java]: Few links for reading
https://www.petrikainulainen.net/software-development/design/why-i-changed-my-mind-about-field-injection/
http://vojtechruzicka.com/field-dependency-injection-considered-harmful/
http://aleksz-programming.blogspot.com/2014/03/why-you-should-autowire-constructor.html

is it clear, or should I explain?

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java (hunk 1 above)
Comment (Malisevych) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java]: Done

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java (hunk 4 above)
Comment (Malisevych) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java]: Done

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java (hunk 3 above)
Comment (Malisevych) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java]: Done

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java (hunk 5 above)
Comment (Malisevych) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/service/ProductPageService.java]: Thanks. Have done

File: common-modules/src/main/java/com/softserve/if072/common/model/validation/product/ValidUnit.java (hunk 10 above)
Comment (Malisevych) [inline common-modules/src/main/java/com/softserve/if072/common/model/validation/product/ValidUnit.java]: Thank you for the review! Remarks took into account

File: MVCapplication/src/main/java/com/softserve/if072/mvcapp/controller/ProductPageController.java (hunk 2 above)
Comment (OlehPochernin) [inline MVCapplication/src/main/java/com/softserve/if072/mvcapp/controller/ProductPageController.java]: Thank you very much! That's quite enough.

**Pull request merged**